File: Core/src/test/java/com/github/ezauton/core/test/simulator/SimulatedMotorTest.java
package com.github.ezauton.core.test.simulator;

import com.github.ezauton.core.action.Action;
import com.github.ezauton.core.action.BaseAction;
import com.github.ezauton.core.actuators.implementations.BaseSimulatedMotor;
import com.github.ezauton.core.simulation.ModernSimulatedClock;
import org.junit.jupiter.api.Test;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class SimulatedMotorTest {
    @Test
    public void testMotor() throws TimeoutException {

        ModernSimulatedClock clock = new ModernSimulatedClock();

        Action action = new BaseAction(() -> {
            BaseSimulatedMotor motor = new BaseSimulatedMotor(clock);

            assertEquals(0, motor.getPosition(), 1E-6);
            motor.runVelocity(1);

            try {
                clock.sleep(1, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                return;
            }

            assertEquals(1, motor.getPosition(), 1E-6);

            try {
                clock.sleep(1, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                return;
            }
            assertEquals(2, motor.getPosition(), 1E-6);
            motor.runVelocity(10);
            assertEquals(2, motor.getPosition(), 1E-6);

            try {
                clock.sleep(1, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                return;
            }
            assertEquals(12, motor.getPosition(), 1E-6);
        });

        clock.add(action);
        clock.runSimulation(5, TimeUnit.SECONDS);

    }
}

File: Core/src/test/kotlin/com/github/ezauton/core/simulator/SimulatedMotorTest.kt
package com.github.ezauton.core.simulator

import com.github.ezauton.conversion.radians
import com.github.ezauton.conversion.sec
import com.github.ezauton.conversion.seconds
import com.github.ezauton.core.action.action
import com.github.ezauton.core.action.delay
import com.github.ezauton.core.action.runWithTimeout
import com.github.ezauton.core.actuators.implementations.BaseSimulatedMotor
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeoutException

class SimulatedMotorTest {
  @Test
  @Throws(TimeoutException::class)
  fun testMotor() = runBlocking {


    val action = action {

      val motor = BaseSimulatedMotor()

      assertEquals(0.0, motor.position.value, 0.1)
      motor.runVelocity(1.0.radians / sec)

      delay(1.seconds)

      assertEquals(1.0, motor.position.value, 0.1)

      delay(1.seconds)

      assertEquals(2.0, motor.position.value, 0.1)
      motor.runVelocity(10.0.radians / sec)
      assertEquals(2.0, motor.position.value, 0.1)

      delay(1.seconds)

      assertEquals(12.0, motor.position.value, 0.1)
    }

    action.runWithTimeout(5.seconds)
  }
}
